irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp

```diff
@@ -78,8 +83,18 @@ EventsExecutor::spin()
   }
   RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
 
-  timers_manager_->start();
-  RCPPUTILS_SCOPE_EXIT(timers_manager_->stop(); );
+  spin_has_finished_.store(false);
+
+  timers_thread_ = std::thread([this](){ timers_manager_->run_timers(); });
+
+  RCPPUTILS_SCOPE_EXIT(
+      timers_manager_->notify_stop();
+
+      if (timers_thread_.joinable()) {
```

Comment:
What about moving this to the destructor, thus removing the `spin_has_finished_` bool and the busy loop in the destructor?

Reply:
Having the thread joining with `timers_thread_.join();` on destructor doesn't guarantee that the `spin()` has finished.
i.e. the call of `RCPPUTILS_SCOPE_EXIT(timers_manager_..)` could end up in seg fault.
Also, if an exception is thrown, the destructor is not called, so `RCPPUTILS_SCOPE_EXIT` should join threads here, right?

Reply:
I don't know; it feels like we are making things way too complex than they should be and all the synchronization seems brittle.

I would expect the multi-threaded executor to have similar problems, but it looks like it doesn't